pageserver/src/tenant_jobs/worker.rs

```diff
@@ -0,0 +1,156 @@
+
+    /// Notified when idle worker should wake up
+    condvar: Arc<Condvar>,
+}
+
+impl<J: Job> Default for Pool<J>
+where
+    J::ErrorType: Debug,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<J: Job> Pool<J>
+where
+    J::ErrorType: Debug,
+{
+    pub fn new() -> Self {
+        Pool {
+            job_table: Mutex::new(JobStatusTable::<J>::new()),
+            condvar: get_shutdown_aware_condvar(),
+        }
+    }
+
+    /// Main loop for a worker. Intended to be spawned via `thread_mgr::spawn`
+    pub fn worker_main(&self, worker_name: String) -> anyhow::Result<()> {
+        let mut job_table = self.job_table.lock().unwrap();
+        while !is_shutdown_requested() {
+            match job_table.take_job(worker_name.clone()) {
+                super::job::TakeResult::Assigned(job) => {
+                    // Run job without holding lock
+                    drop(job_table);
+                    POOL_UTILIZATION_GAUGE
+                        .with_label_values(&["todo_put_pool_name_here"])
+                        .inc();
+                    let result = job.run_safe();
+                    POOL_UTILIZATION_GAUGE
+                        .with_label_values(&["todo_put_pool_name_here"])
+                        .dec();
+                    job_table = self.job_table.lock().unwrap();
+
+                    // Reschedule or report error
+                    job_table.report(job, result);
+                    self.condvar.notify_all();
+                }
+                super::job::TakeResult::WaitUntil(time) => {
+                    let wait_time = time.duration_since(Instant::now());
+                    job_table = self.condvar.wait_timeout(job_table, wait_time).unwrap().0;
+                }
+                super::job::TakeResult::WaitForJobs => {
+                    job_table = self.condvar.wait(job_table).unwrap();
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Schedule a new job
+    pub fn queue_job(&self, job: J) {
+        // TODO what happens if it's already scheduled or stuck?
+        let mut job_table = self.job_table.lock().unwrap();
+        job_table.schedule(job);
+        self.condvar.notify_all();
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{ops::Add, time::Duration};
+
+    use once_cell::sync::OnceCell;
+
+    use super::*;
+    use crate::thread_mgr::{self, ThreadKind};
+
+    #[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+    struct PrintJob {
+        to_print: String,
+    }
+
+    impl Job for PrintJob {
+        type ErrorType = String;
+
+        fn run(&self) -> Result<Option<Instant>, String> {
+            if self.to_print == "pls panic" {
+                panic!("AAA");
+            }
+            println!("{}", self.to_print);
+            Ok(Some(Instant::now().add(Duration::from_millis(10))))
+        }
+    }
+
+    static TEST_POOL: OnceCell<Pool<PrintJob>> = OnceCell::new();
+
+    #[tokio::test]
```

Comment:
Is there any reason to use tokio testing infra? There is no async code involved besides time::sleep which has a non async counterpart in std: https://doc.rust-lang.org/stable/std/thread/fn.sleep.html#

Reply:
no reason. It will go away.